actions/src/openstack_check_actions.py

```diff
@@ -0,0 +1,277 @@
+import ast
+from typing import Callable, Dict
+import logging
+import sys
+from datetime import datetime
+import requests
+from openstack_api.openstack_connection import OpenstackConnection
+from st2common.runners.base_action import Action
+from requests.auth import HTTPBasicAuth
+
+
+class CheckActions(Action):
+
+    # pylint: disable=arguments-differ
+    def run(self, submodule: str, **kwargs):
+        """
+        Dynamically dispatches to the method wanted
+        """
+        func: Callable = getattr(self, submodule)
+        return func(**kwargs)
+
+    # pylint: disable=too-many-arguments, too-many-locals
+    def _check_project_loadbalancers(
+        self, project: str, cloud: str, max_port: int, min_port: int, ip_prefix: str
+    ):
+        """
+        Does the logic for security_groups_check, should not be invoked seperately.
+        """
+
+        with OpenstackConnection(cloud_name=cloud) as conn:
+            sec_group = conn.list_security_groups(filters={"project_id": project})
+        # pylint: disable=too-many-function-args
+
+        bad_rules = self._bad_rules(max_port, min_port, ip_prefix, sec_group)
+        rules_with_issues = self._check_if_applied(bad_rules, cloud, project)
+
+        return rules_with_issues
+
+    @staticmethod
+    def _bad_rules(max_port, min_port, ip_prefix, sec_group):
+        for group in sec_group:
+            for rules in group["security_group_rules"]:
+                # pylint: disable=line-too-long
+                if (
+                    rules["remote_ip_prefix"] == ip_prefix
+                    and rules["port_range_max"] == max_port
+                    and rules["port_range_min"] == min_port
+                ):
+                    return rules
+        return {}
+
+    @staticmethod
+    def _check_if_applied(rules: Dict, cloud, project):
+        rules_with_issues = []
+        with OpenstackConnection(cloud_name=cloud) as conn:
+            servers = conn.list_servers(
+                filters={"all_tenants": True, "project_id": project}
+            )
+
+        for server in servers:
+
+            with OpenstackConnection(cloud_name=cloud) as conn:
+                applied = conn.list_server_security_groups(server.id)
+
+            for serv_groups in applied:
+                if serv_groups["id"] == rules["security_group_id"]:
+                    print("Rule is applied")
+                    rules_with_issues.append(
+                        {
+                            "dataTitle": {"id": server["id"]},
+                            "dataBody": {
+                                "proj_id": project,
+                                "sec_id": serv_groups["id"],
+                                "applied": "Yes",
```

Comment:
Can we pack this into a dataclass, since they all use the same layout in multiple places?

Reply:
Stackstorm requires that payloads be dictionaries. I pass these as part of the payload so this would not work.